wire-ios/Wire-iOS/Sources/UserInterface/Conversation/Message Details/MessageDetailsDataSource.swift

```diff
@@ -157,8 +183,20 @@ final class MessageDetailsDataSource: NSObject, ZMMessageObserver, ZMUserObserve
 
     func userDidChange(_ changeInfo: UserChangeInfo) {
         performChanges {
-            self.reactions = MessageDetailsCellDescription.makeReactionCells(message.sortedLikers)
-            self.readReceipts = MessageDetailsCellDescription.makeReceiptCell(message.sortedReadReceipts)
+            self.reactions = message.usersByReaction.map { reaction, users in
+                MessageDetailsSectionDescription(
+                    headerText: "\(reaction.unicodeValue) \(reaction.displayValue) (\(users.count))",
+                    items: MessageDetailsCellDescription.makeReactionCells(users)
+                )
+            }.filter {
+                !$0.items.isEmpty
+            }
+
+            self.reactions.sort(by: { $0.items.count > $1.items.count })
+
+            self.readReceipts = [
+                MessageDetailsSectionDescription(items: MessageDetailsCellDescription.makeReceiptCell(message.sortedReadReceipts))
+            ]
```

Comment:
This is more duplication which you can avoid if you factor it out.

Reply:
Done!